Add module examples covering macros. (#208)
amzn/ion-docs: src/modules-by-example.adoc

  install_symbols: [local2, local]                    // 'local2' is first
}
----

=== Installing and Using Macros

The local macro table works in essentially the same way as the local symbol table: you import or
define modules that export macros, then you enumerate the modules whose macros you want to
install. The lists of exported macros from each of those modules are concatenated to form a
contiguous address space so that any macro can be referenced by an integer.

> We can now define our macros for two-dimensional geometry with full context:

Reviewer: I don't recall this  context ever being set up in this document so the sentence is really jarring.  What "two-dimensional geometry" are we talking about?  I think you are referring to the by example section, right?

Author: This was an example from the later sections of the macro-by-example document.  I'll rephrase to not assume that dependency.

[{nrm}]
----
*$ion_1_1*
**$ion_encoding**::{
  modules: [
    (*module* geo
      (*macro* point [(int x), (int y)]
        {x: x, y: y})
      (*macro* line  [(point a), (point b)]
        [a, b]))
  ],
  install_macros: [ geo ]
}
(:point 17 28)
(:line (1 2) (3 4))
----

This `geo` module defines macros instead of symbols, using the `*macro*` definition syntax

> explored throughout <<sec:macroexample>>.

Reviewer: Ah, there it is--I think we need to put this above the sentence referring to it.

The `install_macros` field works much like `install_symbols`: it assembles a local macro
table by concatenating the exported macro tables of the referenced modules, which must be
declared within the adjacent `modules` field.

> With macros installed, the document can then invoke them using E-expressions, and the `point` and
> `line` invocations above produce results equivalent to:

Reviewer: *Minor*: Editorial suggestion.

```suggestion
With macros installed, the document can then invoke them using E-Expressions, and the `point`/
`line` invocations above produce results equivalent to:
```

system macros so user-defined macros start at address zero.  In the document above, the first
macro in the first module is `point`, so we could write:

----
(:0 17 28) ⇒ {x:17, y:28}
----

Further, the local macro table tracks the names of installed modules, so that macros can be
addressed using qualified names like `(:geo:point 17 28)`.  Any ambiguity among exported macro names
may be resolved at the point of reference using this syntax.  Qualified addresses work as well,
so `:geo:0` resolves to the macro at address 0 of module `geo`, which is `point`.

All told, Ion text offers four variants of macro references.  Each of these lines is equivalent:

----
(:0         17 28)  (:1        (1 2) (3 4))
(:geo:0     17 28)  (:geo:1    (1 2) (3 4))
(:geo:point 17 28)  (:geo:line (1 2) (3 4))
(:point     17 28)  (:line     (1 2) (3 4))
----

This topic is more interesting when more than one module is involved, so let's table this
for now.

=== Shared Modules

Macros are most useful when they're shared across documents, and for that we use _shared modules_,
a generalization of Ion 1.0's shared symbol tables. As discussed in

> <<_modules_from_the_outside>>, they export both a symbol table and a macro table.

Reviewer: In Ion 1.1, a shared symbol table ***is*** a shared module, right?  Can we just say this explicitly, this text sort of says it, but I feel like this is something to really drill home.

NOTE: We intend to propose a new schema for shared modules, akin to the new `$ion_encoding` schema.
That should be easier to explain and understand than the format below.

For backwards compatibility purposes, shared modules are expressed using the legacy schema for
shared symbol tables, adding a `module` field to hold macro definitions:

[{nrm}]
----
*$ion_1_0*
**$ion_shared_symbol_table**::{
  name: "com.example.graphics.3d",
  version: 1,
  symbols: ["x", "y", "z"],

  // Schema addition follows:
  module: $ion_1_1::(
    (*macro* point [(int x), (int y), (int z)]
      {x: x, y: y, z: z})
    (*macro* line  [(point a), (point b)]
      [a, b])
    (*macro* poly  [(point first), (point... rest)]
      [first, rest])
  )
}
----

The `module` field here is very similar to the `*module*` S-expression inside `$ion_encoding`.
Here, no symbolic name is declared, since one will be assigned when the module is ``*import*``ed.

> No `*symbols*` clause is allowed, since those are expected to be in the legacy `symbols` field.

Reviewer: Ugh, I know why this is here, but it is annoying that we have to deal with different parsing contexts and different rules for this stuff.

Author: I tried really hard to minimize any differences, this is the best I've been able to come up with.

I considered allowing a `symbols` clause inside the module, assuming there's nothing in the legacy `symbols` field (and thus no need to combine them), but then the overall SST isn't 1.0-meaningful.

For comparison, here's a functionally-equivalent local definition:

[{nrm}]
----
**$ion_encoding**::{
  modules: [
    (*module* g3d
      (*symbols* ["x", "y", "z"])
      (*macro* point [(int x), (int y), (int z)]
        {x: x, y: y, z: z})
      (*macro* line  [(point a), (point b)]
        [a, b])
      (*macro* tri   [(point a), (point b), (point c)]
        [a, b, c]))
----

The `$ion_shared_symbol_table` document above is encoded in Ion 1.0 format, despite containing
information that only applies to an Ion 1.1 implementation.  Shared symbol tables are
communicated via the Ion data model, which is guaranteed consistent across all Ion 1.x
specifications, so encoding modules can be expressed using any Ion version with no change in
semantics.  To accomplish this, we require the IVM-like `$ion_1_1` annotation on the `module`
field, denoting the <<spec-version,spec version>> that provides meaning to the module.

=== Using Shared Macros

With a shared module at hand, we can import it and install its macros:

[{nrm}]
----
*$ion_1_1*
**$ion_encoding**::{
  modules: [

>     (*import* g3d "com.example.graphics.3d" 1),     // Import it

Reviewer: *Side Note*: I still dislike that if my shared module was called `g3d` that I would have to write something like:
```
(import g3d "g3d" 1)
```

Author: I've cut #213 to discuss.

    (*module* geo
      (*macro* point [(int x), (int y)]
        {x: x, y: y})
      (*macro* line  [(point a), (point b)]
        [a, b]))
  ],
  install_macros: [ geo, g3d ]                    // Install it
}
----

We now have a problem: the names `point` and `line` are ambiguous, referring to two different
macros each.  Thankfully, we can use qualified references to disambiguate:

----
(:geo:point 17 28)  (:g3d:point 20 18 45)
(:geo:0     17 28)  (:g3d:0     20 18 45) // Equivalent
----

In fact, we _must_ do so.  An E-expression with an un unqualified macro name is erroneous when
the name is ambiguous, meaning that two installed modules map it to different macros.

[{nrm}]
----
(:point 17 28) ⇒ **error**: ':point' is ambiguous, exported by 'geo' and 'g3d'.
----

> Another thing to note in this directive is that the `**import** g3d` declaration includes a symbol

Reviewer: *Minor*: Editorial suggestion.

```suggestion
Another thing to note in the directive used above is that the `**import** g3d` declaration includes a symbol
```

> table name and version, but no max_id argument.  As with legacy imports, absence of max_id forces

Reviewer: What is a "legacy import?" is that something that we define somewhere or do we mean  an import in an Ion 1.0 local symbol table?

Author: I mean the latter, so I'll just say that.

the Ion implementation to acquire the symbol table entity with exactly the stated version.  While
this is generally not best-practice for importing symbols, exact-match is a precondition for

> using the module in `install_macros`.  In other words, when a document is encoded using macros,

Reviewer: *Minor*: Editorial suggestion.  I think we should be clearer on the requirement.

```suggestion
this is generally not best-practice for importing symbols, exact-match is a **requirement** for
using the module in `install_macros`.  In other words, when a document is encoded using macros,
```

symbolic name to the module for qualified references.

We know how to use the macro:

[{nrm}]
----
*$ion_1_1*
**$ion_encoding**::{
  modules: [(*import* chart "com.example.charts" 1)],
  install_macros: [chart]
}
(:scatterplot (3 17) (395 23) (15 48) (2023 5))
----

While the signature of `point` is now implicit in the signature of `scatterplot`, and while the
macro expander will invoke `point` while expanding `scatterplot`, neither `point` nor
the module containing it is in scope within the document:

[{nrm}]
----
(:point 25 10)   ⇒ **error**: no installed module exports a macro named 'point'.
(:geo:point 2 1) ⇒ **error**: no module named 'geo' is installed.
----

In particular, `geo` is not in the encoding environment's available modules, since it wasn't
imported into it:

[{nrm}]
----
*$ion_1_1*
**$ion_encoding**::{
  modules: [(*import* chart "com.example.charts" 1)],
  install_macros: [chart, geo]
}
  ⇒ **error**: no module named 'geo' is available for installation.
----

When the Ion implementation loads the 'chart' module, it will transitively load the geometry
module as well, but the import of `com.example.geometry` by `com.example.charts` is

> _private_ to the importer.

Reviewer: *Minor*: Editorial suggestion.

```suggestion
_not visible by name_ to the importer.
```

is to give new names to existing macros.  Suppose we want to add a 3d chart to our module, so we
import both the 2d and 3d modules:

[{nrm}]
----
*$ion_1_1*
**$ion_encoding**::{
  modules: [
    (*module* chart
      (*import* geo "com.example.geometry" 1)
      (*import* g3d "com.example.graphics.3d" 1)
      (*macro* scatterplot [(point\... points)]

  ⇒ **error**: 'point' is ambiguous, exported by 'geo' and 'g3d'.
----

The most direct way to fix this is to use a qualified reference.  We've seen this used in
E-expressions like `(:geo:point 17 28)`, but now we need it in a signature where the special
smile syntax does not apply.  Instead, use a quoted symbol:

[{nrm}]
----
      (*macro* scatterplot [(':geo:point' \... points)]
        [points]))
----

That has the intended effect of keeping `scatterplot` using 2D points, but it's somewhat awkward.
A more ergonomic approach is to introduce an alias to disambugate:

[{nrm}]
----
  modules: [
    (*module* chart
      (*import* geo "com.example.geometry" 1)
      (*import* g3d "com.example.graphics.3d" 1)
      (*alias* point2 ':geo:point')               // <1>

>       (*macro* scatterplot [(point2 \... points)]  // <2>

Reviewer: I notice the clever use of `point2` to avoid shadowed  names.  Is this where `(':point2' ...)` could be used in the macro definition?

Also even with the shadowing, couldn't this lexical alias override the ambiguity?

I think this is worth showing as examples, because I am not sure I know what the rules are.

Author: Yes, you could write `(':point2' \... points)` in the signature, since that colon forces resolution among macro/alias declarations in this module.

But I kinda want to avoid getting into all the rules here.

PS I didn't think this was clever, its kind a the point of the example...

        [points])
      ...
----

<1> Declaration of alias `point2`.
<2> Use of that new name in a signature.

Aliases can only be declared within a module, where they can be used wherever a macro reference
occurs, including for macro invocations in the template language.  In addition to disambiguation,
they can be used to shorten long names, or to give names to anonymous macros.

=== Exports

Unlike `macro` definitions, aliases are not automatically exported from the module where they are
declared; they are presumed to be implementation details.  Sometimes it's helpful to make them
available to consumers of the module, and for that they can be exported:

[{nrm}]
----
*$ion_1_1*
**$ion_encoding**::{
  modules: [
    (*import* geo "com.example.geometry" 1),
    (*import* g3d "com.example.graphics.3d" 1),
    (*module* local
      (*import* geo)
      (*import* g3d)
      (*alias* point2 ':geo:point')
      (*alias* point3 ':g3d:point')
      (*export* point2 point3))
  ],
  install_macros: [ local, geo, g3d ]
}
(:point2 93 5)
(:point3 0 12 33)
----

Exports can also be used to "pass through" selected macros from an imported module: `(*export*

> ':g2d:line')` exports the name `line` from the enclosing module.

Reviewer: I think this is different than:
```
(alias line ':g2d:line')
(export line)
```

And I think the difference is I don't get a local name, but is that right (and this question makes me think we should have text that illustrates if this is a difference or not).

Author: You are correct. It has the same effect to the module's consumer, but there's no local name.  I'll add the example.

IMPORTANT: The macro names exported by a module must be unique, regardless of whether they are
exported implicitly via `*macro*` or explicitly via `*export*`.

=== Extending the Macro Table

Some Ion use cases benefit from defining macros "on the fly" in response to repeated content.
The techniques we used to extend the symbol table in <<_extending_the_symbol_table>> work for
the macro table as well:

[{nrm}]
----
*$ion_1_1*
**$ion_encoding**::{
  modules: [(*module* local
              (*symbols* ["s1", "s2"])
              (*macro* m1 ...))],
  install_symbols: [local]
  install_macros:  [local]
}

// ... application data ...

**$ion_encoding**::{
  modules: [(*retain* *{asterisk}*),
            (*module* local2
              (*symbols* /{asterisk} new symbols {asterisk}/)
              (*macro* m2 ...))
  ],
  install_symbols: [local, local2]
  install_macros:  [local, local2]
}
----

=== Separate Installation

The preceding example has some repetition between `install_symbols` and `install_macros`,

> illustrating that the symbol and macro tables are maintained independently.  There's no

Reviewer: Maybe have an example where:

```
  install_symbols: [local, local2]
  install_macros:  [local2, local]
```

  {
    install_symbols: [module_names],
    install_macros:  [module_names]
  })
----

Invoked as:

[{nrm}]
----
**$ion_encoding**::{
  modules: [(*import* foo ...),
            (*import* bar ...),
            (*import* baz ...)],
  (:install_both bar foo baz)
}
----

This leverages <<_splicing_in_encoded_data,splicing>> to add two fields to the enclosing struct.

=== Prioritization

The features we've explored can be combined to achieve fine-grained control over the allocation
of macro and symbol addresses.  This lets document authors assign the smallest opcodes to the
most used macros and symbols.

Let's assume that our graphics modules have grown to include a large number of
macros, far more than the 64 that can be invoked with a single-byte opcode.  If we know that our
document invokes, say, 3D `point` and `tri` more than anything else, we can grant them
single-byte opcodes by ensuring they show up first among the installed macros:

[{nrm}]
----
*$ion_1_1*
**$ion_encoding**::{
  modules: [
    (*import* geo "com.example.geometry" 1),
    (*import* g3d "com.example.graphics.3d" 1),
    (*module* priority
      (*import* g3d)
      (*export* point tri))
  ],

>   install_macros: [priority, geo, g3d]

Reviewer: I think there needs to be text around the scoping of `import`, it surprised me that `(import g3d)` was needed in the locally defined module since you had imported it in the `modules` section.

Author: TBH I've wavered on the necessity of this.  My last doc round eleminited the requirement, and the code in a module could see all macros exported from preceding `import` and `module` declarations.  But I think that will lead to more name collision and scoping confusion, and change-at-a-distance, with no way to "opt out" of that visibility.  In short, more qualified names and/or aliases just to workaround having things visible that you don't need or want.

When I wrote up this example I considered adding a new keyword for this mechanism, since this isn't really the same kind of "import".  My strawman would be perhaps `(use <module-name>+)`.   Would a different syntax make more sense?

Update: you could then `(use *)` to get the "everything is visible" state.
